### ISSUE TITLE
Definitional classes should support `Cumulative`

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#18147]
Consider this code:
```coq
Set Universe Polymorphism.
Unset Universe Minimization ToSet.
Set Polymorphic Inductive Cumulativity.
(*Cumulative*) Class A@{i} (T : Type@{i}) := a : T.
(*Error: Definitional classes do not support the inductive cumulativity attribute.*)
(*Cumulative*) Class A'@{i} (T : Type@{i}) := a' : T.
Class B@{i} (T : Type@{i}) := mkB { b : T }.
Class M (T : Type -> Type) := m : False.
Instance Anat@{} : A@{Set} nat := O.
Instance Bnat@{} : B@{Set} nat := @mkB@{Set} _ O.
Declare Instance MA : M@{Set Set} A@{Set}.
Definition Bnat2@{u} : B@{u} nat := _.
Fail Definition Anat2@{u} : A@{u} nat := _.
(*
The command has indeed failed with message:
The following term contains unresolved implicit arguments:
  ?a
More precisely:
- ?a: Cannot infer this placeholder of type "A nat" (no type class instance found).
 *)
Fail Definition MA2@{u} : M A@{u} := _.
Fail Definition MA'2@{u} : M A'@{u} := _.
#[export] Typeclasses Transparent A A' M.
Definition Anat2@{u} : A@{u} nat := _.
Fail Definition A'nat2@{u} : A'@{u} nat := _. (* good *)
Fail Definition MA'2@{u} : M A'@{u} := _. (* but I didn't want to convert A' with A :-( *)
```
I want to be able to apply `Cumulative` to definitional classes (and constants).  That is, I want to tell Coq: when facing a conversion problem `C@{i j ...} a b ... =?= C@{i' j' ...} a' b' ...`, do not unfold `C` (the arguments must be convertible).  However, when facing a conversion problem of the form `C@{i j ...} =?= C@{i' j' ...}` where the universes do not unify, then unfolding `C` is permissible.  (I guess this is really a request for something like `Hint Cumulative` / `Typeclasses Cumulative`, which is to be a restricted version of `Hint Transparent` / `Typeclasses Transparent`.)

**Comment by SkySkimmer** [comment]
Definitional classes are just definitions so should this be "definitions should support cumulative"?

**Comment by JasonGross** [comment]
No, I think there should be `Hint Cumulative`, and then `Cumulative Class` should mean `Hint Cumulative ... : typeclass_instances`.  For most definitions, it's not clear what database to insert the hint in (and I think picking `typeclass_instances` for all definitions would be poor / confusing design.)  (Plausibly we should also have attributes for `Hint Opaque` / `Hint Transparent` at class declaration time.)

**Comment by SkySkimmer** [comment]
We for sure don't want to overload Cumulative Class that way.

I'm also not convnced that this Hint Cumulative mechanism is a good idea and can be reasonably implemented.

**Comment by JasonGross** [comment]
> I'm also not convnced that this Hint Cumulative mechanism is a good idea and can be reasonably implemented.

Can you elaborate more?

And/or, what about this alternative proposal?

Fully parameterize the constant unfolding heuristic in conversion on a table of "LHS head constant -> RHS head constant -> left|right|both|neither".
Shortcuts to "when seeing `C1 args1 =?= C2 args2` and preparing to unfold one of `C1` or `C2`, unfold left|right|both|neither":
- `Hint Strategy n [C]` means "unfold `C` on either side before any constants with higher strategy and after any constants with lower strategy, default level is 0, if the strategy is level is the same for two constants fall back to the current LTR/RTL/whatever heuristic
- `Hint Opaque C` = `Hint Strategy opaque [C]` = `Hint Strategy ∞ [C]`
- `Hint Transparent C` = `Hint Strategy 0 [C]`
-  `Hint Strategy expand [C]` = `Hint Strategy -∞ [C]`
- `Hint Equivalent Keys C1 C2` = add `C1 -> C2 -> (default direction)` and `C2 -> C1 -> (default direction)` and `C1 -> C1 -> both` and `C2 -> C2 -> both`
- `Hint Cumulative C` = `Declare Equivalent Keys C` aka unfold unfold `C` when it's at the head on both sides

**Comment by JasonGross** [comment]
Okay, I am backtracking here, I think `Definition`s in general should support cumulativity and variance, given that the above does not interact adequately with cumulative inductives.  Let me pull out a separate issue for this, though.

**Comment by JasonGross** [comment]
I have found an example where we want cumulatively for definitions that isn't handled by fine-grained unfolding, but I will make a new issue.
